django/db/models/functions/json.py

```diff
@@ -0,0 +1,258 @@
+from django.db import NotSupportedError
+from django.db.models.constants import LOOKUP_SEP
+from django.db.models.expressions import Func, Value
+from django.db.models.fields.json import compile_json_path
+from django.db.models.functions import Cast
+
+
+class JSONSet(Func):
+    def __init__(self, expression, output_field=None, **fields):
+        if not fields:
+            raise TypeError("JSONSet requires at least one key-value pair to be set.")
+        self.fields = fields
+        super().__init__(expression, output_field=output_field)
+
+    def _get_repr_options(self):
+        return {**super().get_repr_options(), **self.fields}
+
+    def resolve_expression(
+        self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False
+    ):
+        c = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        # Resolve expressions in the JSON update values.
+        c.fields = {
+            key: (
+                value.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+                if hasattr(value, "resolve_expression")
+                else value
+            )
+            for key, value in self.fields.items()
+        }
+        return c
+
+    def as_sql(
+        self,
+        compiler,
+        connection,
+        function=None,
+        template=None,
+        arg_joiner=None,
+        **extra_context,
+    ):
+        if not connection.features.supports_partial_json_update:
+            raise NotSupportedError(
+                "JSONSet() is not supported on this database backend."
+            )
+        copy = self.copy()
+        new_source_expressions = copy.get_source_expressions()
+
+        for key, value in self.fields.items():
+            key_paths = key.split(LOOKUP_SEP)
+            key_paths_join = compile_json_path(key_paths)
+            new_source_expressions.append(Value(key_paths_join))
```

Comment:
I am not sure how this can be a SQL injection vector. The `compile_json_path()` function uses `json.dumps()` for each key in the path, so it will escape quotes and backslashes properly. Then, we use `Value` which will separate the expression into the SQL template and parameters in its `as_sql()`. 

I added some tests in https://github.com/django/django/pull/18758/commits/08f07dc8adc701f55a85f78f146ff51251397f9d. Although, I just noticed it seems SQLite cannot handle double quotes in the path. See also https://stackoverflow.com/questions/67993982. I think we can address this with a feature flag.

However, if there is a better way to do it, please let me know. Thanks!

Reply:
You're right about the key JSON quoting in this case, I thought it was the same thing as in the `as_postgres` method.

The issue with double quotes on SQLite and likely with single quotes on Oracle in generalized in ticket-32213 and ticket-35842.